### PR TITLE
Unit test

### PR DESCRIPTION
**Pull request by davidos-b**

### CODE REVIEW

File: modules/ardrivo/test/String.cxx [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+TEST_CASE("Tests whether or not a String starts with the characters of another String" "[startsWith]")
+{
+	String test = "test";
+	String test2 = "test";
+	String test3 = "est";
+	String test4 = "test";	
+
+	REQUIRE(test.startsWith(test2) == true);
```
Comment (RuthgerD) [inline modules/ardrivo/test/String.cxx:21]: These can be simplified by omitting the == true (and for false just do `!`)

File: modules/ardrivo/test/String.cxx [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+	String test3 = "est";
+	String test4 = "test";
+
+	REQUIRE(test.startsWith(test2) == true);
+	REQUIRE(test3.startsWith(test4) == false);
+}
+
+TEST_CASE("Copies the String�s characters to the supplied buffer" "[getBytes]")
+{
+	using Catch::Contains;
```
Comment (RuthgerD) [inline modules/ardrivo/test/String.cxx:38]: Please dont do 'using ..some namespace..' when you only use it once

File: modules/ardrivo/test/String.cxx [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+TEST_CASE("Check operators" "[String::operator]")
+{
+	String test = "d";
+	String test2 = "AB";
+
+	SECTION("Operator ==")
+	{
+		REQUIRE(test.operator==(test2) == false);
```
Comment (RuthgerD) [inline modules/ardrivo/test/String.cxx:200]: Why do you call the operator directly? also == false should be omitted, just use `!`.

File: modules/ardrivo/test/String.cxx (hunk 2 above)
Comment (davidos-b) [reply modules/ardrivo/test/String.cxx:38]: Well it is not used only once? Or what do you mean?

Comment (AeroStun) [reply modules/ardrivo/test/String.cxx:38]: This isn't a `using namespace`; this is equivalent to  `using Contains = Catch::Contains;`

File: modules/ardrivo/test/Arduino.cxx [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    clock_t start = clock();
+    delay(1000);
+    clock_t stop = clock();
+    double elapsed = (double)(stop - start) / CLOCKS_PER_SEC;
+
+    //Added some margin
+    REQUIRE(elapsed >= 0);
+    REQUIRE(elapsed < 1.1);
+}
+TEST_CASE("delay the thread in microseconds", "[delayMicroseconds]") {
```
[the remark this answers is not in the document]
Comment (AeroStun) [reply modules/ardrivo/test/Arduino.cxx:100]: `std::chrono::steady_clock`

File: modules/ardrivo/test/Arduino.cxx [hunk 5]
```diff
@@ -87,25 +88,29 @@ TEST_CASE("Set tone frequency on pin", "[tone], [noTone]")
...
-    double elapsed = (double)(stop - start) / CLOCKS_PER_SEC;
-
+    const auto end = std::chrono::steady_clock::now();
+    const auto duration = std::chrono::duration_cast<std::chrono::seconds>(end - start);
     //Added some margin
-    REQUIRE(elapsed >= 0);
-    REQUIRE(elapsed < 1.1);
+
+    REQUIRE(duration > 0s);
+    REQUIRE(duration < 1.5s);
```
Comment (AeroStun) [inline modules/ardrivo/test/Arduino.cxx:100]: Your margin allows -99% and +49%; that's a statistically terrible test

File: modules/ardrivo/test/Arduino.cxx [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+
+TEST_CASE("delay the thread in microseconds", "[delayMicroseconds]") 
+{
+    const auto start = std::chrono::steady_clock::now();
+    delay(1000);
```
Comment (AeroStun) [inline modules/ardrivo/test/Arduino.cxx:107]: You're effectively testing `delay` but your test says you should be testing `delayMicroseconds`

File: modules/ardrivo/test/Arduino.cxx [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+    const auto end = std::chrono::steady_clock::now();
+    const auto duration = std::chrono::duration_cast< std::chrono::microseconds >(end - start);
+    //Added some margin
+
+    REQUIRE(duration > 0us);
+    REQUIRE(duration < 1001000us);
```
Comment (AeroStun) [inline modules/ardrivo/test/Arduino.cxx:113]: Again, terribly large statistical interval

File: modules/ardrivo/test/Arduino.cxx [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+TEST_CASE("Return number of microseconds since arduino board began", "[micros]") {
+    init_fake();
+    const unsigned long duration = micros();
+
+    //Added some margin
+    REQUIRE(duration >= 15); 
+    REQUIRE(duration < 70);
```
Comment (AeroStun) [inline modules/ardrivo/test/Arduino.cxx:122]: Margin is useless; Ardrivo tests are unordered and take random amounts of time

File: modules/ardrivo/test/Arduino.cxx [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+TEST_CASE("Return number of milliseconds since arduino board began", "[millis]") {
+    init_fake();
+    const unsigned long duration = millis();
+
+    //Added some margin
+    REQUIRE(duration >= 0);
+    REQUIRE(duration < 30);
```
Comment (AeroStun) [inline modules/ardrivo/test/Arduino.cxx:131]: Same as above

File: modules/ardrivo/test/Arduino.cxx [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+    REQUIRE(isAlphaNumeric('1') == true);
+    REQUIRE(isAlphaNumeric('!') == false);
+}
+
+TEST_CASE("Is the character a ascii", "[isAscii]") {
+    // REQUIRE(isAscii('a') == 97); //Eeehhh??????????????
+
+    int var = 97;
+    //std::cout << (char)var;
+}
```
Comment (AeroStun) [inline modules/ardrivo/test/Arduino.cxx:166]: Missing test

File: modules/ardrivo/test/Arduino.cxx [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+    REQUIRE(isUpperCase('A') == true);
+    REQUIRE(isUpperCase('a') == false);
+}
+
+TEST_CASE("Checks if a character is a blank character" "[isWhiteSpace]") 
+{
+    REQUIRE(isWhitespace(' ') == true);
+    REQUIRE(isWhitespace('\t') == true);
+    REQUIRE(isWhitespace('a') == false);
+}
```
Comment (AeroStun) [inline modules/ardrivo/test/Arduino.cxx:224]: Spurious `bool -> bool` comparisons

File: modules/ardrivo/test/Arduino.cxx [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+    int num = std::rand();
+    int num2 = std::rand();
+
+    std::srand(2);
+    int num3 = std::rand();
+
+    REQUIRE(num != num2);
+    REQUIRE(num == num3);
+}
+
```
Comment (AeroStun) [inline modules/ardrivo/test/Arduino.cxx:244]: The test is lying, since it is testing C random function rather than the Arduino ones

File: modules/ardrivo/test/Arduino.cxx [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+void fun() {}
```
Comment (AeroStun) [inline modules/ardrivo/test/Arduino.cxx:245]: Is it fun?

File: modules/ardrivo/test/Arduino.cxx [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+
+TEST_CASE("Check if vector contains interupt", "[attachInterrupt]") 
+{
+    // First test
+    void (*user_func)() = &fun;
+    attachInterrupt(0, (*user_func), 1);
+    auto test = board_data->interrupts_handlers[0];
+    REQUIRE(test.second == 1);
```
Comment (AeroStun) [inline modules/ardrivo/test/Arduino.cxx:253]: Should also check that the handler is correctly saved

File: modules/ardrivo/test/Arduino.cxx [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+TEST_CASE("Delete a interupt from vector", "[detachInterrupt]") 
+{
+    auto test = board_data->interrupts_handlers[0];
+    REQUIRE(test.first != NULL);
+    board_data->interrupts_handlers.erase(board_data->interrupts_handlers.begin());
+    test = board_data->interrupts_handlers[0];
+    REQUIRE(test.first == NULL);
+}
```
Comment (AeroStun) [inline modules/ardrivo/test/Arduino.cxx:268]: This test does not even call the very function it is supposed to test

File: modules/ardrivo/test/HardwareSerial.cxx [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+    ranges::for_each(loc_board_data->uart_buses, [] (DynaBufferBus& bus)
+    {
+        bus.rx.resize(3);
+        bus.tx.resize(3);
+    });
+
+    init(loc_board_data.release(), loc_board_info.release());
+}
+
+TEST_CASE("Hardwareserail begin and write" "[begin], [write]")
```
Comment (AeroStun) [inline modules/ardrivo/test/HardwareSerial.cxx:40]: Typo

File: modules/ardrivo/test/String.cxx [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+
+	SECTION("Operator >=")
+	{
+		REQUIRE(test >= test2);
+	}
+
+	SECTION("Operator+")
+	{
+		std::string str = "Hello, ";
+		str.operator+=("world");
```
Comment (AeroStun) [inline modules/ardrivo/test/String.cxx:222]: This is also more commonly written `str += "world"`

File: modules/ardrivo/test/String.cxx [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+	}
+
+	SECTION("Operator <")
+	{
+		REQUIRE(!(test < (test2)));
+	}
+
+	SECTION("Operator <=")
+	{
+		REQUIRE(!(test <= (test2)));
```
Comment (AeroStun) [inline modules/ardrivo/test/String.cxx:206]: Please remove spurious parens on second operands

File: modules/ardrivo/test/String.cxx [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+
+TEST_CASE("Converts a valid String to an float. The input String should start with an float number" "[toFloat]")
+{
+	String test = "1.0123 and 23";
+	float num =  test.toFloat();
+	REQUIRE(num == 1.0123f);
+	String test2 = "123.23545";
+	float num2 =  test2.toFloat();
+	REQUIRE(num2 == 123.23545f);
+}
```
Comment (AeroStun) [inline modules/ardrivo/test/String.cxx:139]: Dangerous tests, since you have not proven that these values can be represented by all standard compliant floating-point implementations; do range checks instead

File: modules/ardrivo/test/String.cxx [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+	REQUIRE_THAT(arr, Contains("st"));
+}
+
+TEST_CASE("Copies the String�s characters to the supplied buffer" "[toCharArray]")
+{
+	String test = "test";
+	char arr[5];
+	test.toCharArray(arr, test.length());
+	REQUIRE_THAT(arr, Contains("test"));
+}
```
Comment (AeroStun) [inline modules/ardrivo/test/String.cxx:109]: UB test since `arr[4]`is uninitialized

File: modules/ardrivo/test/String.cxx [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+	REQUIRE_THAT(arr, EndsWith( "cd" ));
+}
+
+TEST_CASE("Test if string is given a new capacity" "[reserve]")
+{
+	std::string test = "test is test";
+	REQUIRE(test.capacity() == 15);
+	test.reserve(50);
+	REQUIRE(test.capacity() == 63);
+}
```
Comment (AeroStun) [inline modules/ardrivo/test/String.cxx:83]: Why are you testing `std::string`? your goal should be to test `String`

File: modules/ardrivo/test/String.cxx [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+
+TEST_CASE("Locates a character or String within another String" "[indexOf]")
+{
+	String test = "test";
+	String test2 = "test";
+	String test3 = "est";
+
+	REQUIRE(!test.indexOf(test2));
+	REQUIRE(test.indexOf(test3));
```
Comment (AeroStun) [inline modules/ardrivo/test/String.cxx:55]: `.indexOf` returns an integer which should probably be checked for exactness rather than narrowed to a `bool`

File: modules/ardrivo/test/String.cxx [hunk 23]
```diff
@@ -0,0 +1,227 @@
...
+using Catch::EndsWith; 
+
+TEST_CASE("Test compare a string to another" "[compareTo]")
+{
+	String test = "test";
+	String test2 = "test";
+	String test3 = "test";
+	String test4 = "est";	
+	REQUIRE(!test.compareTo(test2));
+	REQUIRE(test3.compareTo(test4));
```
Comment (AeroStun) [inline modules/ardrivo/test/String.cxx:14]: `String::compareTo` is a three-way comparison; you are only testing for equality/inequality; insufficient test

File: modules/ardrivo/test/HardwareSerial.cxx [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+{
+    HardwareSerial test;
+    REQUIRE(test.availableForWrite() == 2147483647);
+}
+
+TEST_CASE("Check peek of the next byte" "[peek]")
+{
+    HardwareSerial test;
+    test.begin(1);
+    REQUIRE(test.peek() == 150);
```
Comment (AeroStun) [inline modules/ardrivo/test/HardwareSerial.cxx:78]: Tests are not run in order; this test may fail

File: modules/ardrivo/test/HardwareSerial.cxx [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+    test.begin(1);
+    REQUIRE(test.peek() == 150);   
+}
+
+TEST_CASE("The first byte" "[read]")
+{
+    HardwareSerial test;
+    test.begin(1);
+    REQUIRE(test.read() == 150);
+    REQUIRE(test.read() == 0);
```
Comment (AeroStun) [inline modules/ardrivo/test/HardwareSerial.cxx:86]: Tests are not run in order; this test may fail

File: modules/ardrivo/test/HardwareSerial.cxx [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+
+    REQUIRE(static_cast<int>(a) == 150);
+    REQUIRE(static_cast<int>(b) == 150);
+    REQUIRE(static_cast<int>(c) == 0);
+}
+
+TEST_CASE("Check available bytes for reading" "[available]")
+{
+    HardwareSerial test;
+    REQUIRE(test.available() == 3);
```
Comment (AeroStun) [inline modules/ardrivo/test/HardwareSerial.cxx:65]: Tests are not run in order; this test may fail

File: modules/ardrivo/test/Arduino.cxx [hunk 27]
```diff
@@ -0,0 +1,268 @@
...
+    REQUIRE(num == num3);
+}
+
+void fun() {}
+
+TEST_CASE("Check if vector contains interupt", "[attachInterrupt]") 
+{
+    // First test
+    void (*user_func)() = &fun;
+    attachInterrupt(0, (*user_func), 1);
```
Comment (AeroStun) [inline modules/ardrivo/test/Arduino.cxx:251]: Do not dereference a function pointer; the confusion induced by implicit conversion from function references to function pointers is not be sought

File: modules/ardrivo/test/Arduino.cxx [hunk 28]
```diff
@@ -0,0 +1,268 @@
...
+        auto i = GENERATE(0, 1, 2);
+        digitalWrite(i, HIGH);
+        REQUIRE(digitalRead(i) == HIGH);     
+    }
+    SECTION("Set analog pins") {
+        analogWrite(1, 0);
+        REQUIRE(analogRead(1) == 0);
+        analogWrite(1, 500);
+        REQUIRE(analogRead(1) == 500);
+    }
```
Comment (AeroStun) [inline modules/ardrivo/test/Arduino.cxx:65]: I am curious as to how you manage to have a pin in both input and output mode at the same time

File: modules/ardrivo/test/Arduino.cxx [hunk 29]
```diff
@@ -0,0 +1,268 @@
...
+    REQUIRE(num != num2);
+    REQUIRE(num == num3);
+}
+
+void fun() {}
+
+TEST_CASE("Check if vector contains interupt", "[attachInterrupt]") 
+{
+    // First test
+    void (*user_func)() = &fun;
```
Comment (AeroStun) [inline modules/ardrivo/test/Arduino.cxx:250]: Good place for `auto`

File: modules/ardrivo/test/Arduino.cxx [hunk 30]
```diff
@@ -0,0 +1,322 @@
...
+
+    for (int i = 0; i < 100; i++)
+    {
+        arr3[i] = random(100);
+    }
+
+    for (int i = 0; i < 100; i++)
+    {
+        REQUIRE(arr[i] == arr3[i]);
+    }
```
Comment (RuthgerD) [inline modules/ardrivo/test/Arduino.cxx:289]: pretty sure you can just `arr == arr3` (on std::array and std::vector) otherwise do `ranges::equal` (<range/v3/algorithm/equal.hpp> iirc) or [`std::equal`](https://en.cppreference.com/w/cpp/algorithm/equal)

File: modules/ardrivo/test/Arduino.cxx [hunk 31]
```diff
@@ -0,0 +1,322 @@
...
+    long arr[100];
+    long arr2[100];
+    long arr3[100];
+    auto count = 0;
+    randomSeed(2);
+
+    for (int i = 0; i < 100; i++)
+    {
+        arr[i] = random(100);
+    }
```
Comment (RuthgerD) [inline modules/ardrivo/test/Arduino.cxx:262]: use[ `std::generate`](https://en.cppreference.com/w/cpp/algorithm/generate) (or ranges::generate) and maybe put this in a lambda because this code is copy pasted 3 times
e.g.
```cpp
auto gen_random = [](auto& arr){ ranges::generate(arr, [](auto& el){ return random(100); }); };
```

Just try to not dupe code by using standard algorithm functions :+1:

File: modules/ardrivo/test/HardwareSerial.cxx [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+    test.begin(1);
+    int succes = test.write(150);
+    REQUIRE(succes == 1);
+    std::byte a = board_data->uart_buses[0].rx[0];
+    std::byte b = board_data->uart_buses[2].rx[2];
+    std::byte c = board_data->uart_buses[1].rx[1];    
+
+    REQUIRE(static_cast<int>(a) == 150);
+    REQUIRE(static_cast<int>(b) == 150);
+    REQUIRE(static_cast<int>(c) == 0);
```
Comment (RuthgerD) [inline modules/ardrivo/test/HardwareSerial.cxx:59]: can all be inlined

File: modules/ardrivo/test/String.cxx (hunk 2 above)
Comment (RuthgerD) [reply modules/ardrivo/test/String.cxx:43]: Oh my bad didnt read it right but still think its better to remove the `using` and just do `Catch::Contains("test")` the single time its used

File: modules/ardrivo/test/String.cxx [hunk 33]
```diff
@@ -0,0 +1,241 @@
...
+}
+
+TEST_CASE("Get a substring of a String" "[subsring]")
+{	
+	using Catch::Contains; 
+	String test = "test";
+	String newString = test.substring(2);
+	char arr[3];
+	strcpy(arr, newString.c_str());
+	REQUIRE_THAT(arr, Contains("st"));
```
Comment (RuthgerD) [inline modules/ardrivo/test/String.cxx:108]: cant you directly do `REQUIRE_THAT(newString.c_str(), Catch::Constrains("st"));`? (seen it multiple times)

File: modules/ardrivo/test/String.cxx [hunk 34]
```diff
@@ -0,0 +1,241 @@
...
+	String test = "test";
+	char arr[5] = "init";
+	test.toCharArray(arr, test.length());
+	REQUIRE_THAT(arr, Contains("test"));
+}
+
+TEST_CASE("Converts a valid String to an integer. The input String should start with an integer number" "[toInt]") 
+{
+	String test = "1 and 23";
+	unsigned long num =  test.toInt();
```
Comment (RuthgerD) [inline modules/ardrivo/test/String.cxx:123]: good place to just use `auto` but I'm pretty sure toInt() returns `long` and not `unsigned long`

File: include/BoardInfo.hxx [hunk 35]
```diff
@@ -24,7 +24,7 @@
 struct PinCapability {
     bool digital_in : 1;
     bool digital_out : 1;
-    bool analog_in : 1;
+    bool analog_in : 1  ;
```
Comment (RuthgerD) [inline include/BoardInfo.hxx:27]: Don't touch other files besides those in /test please (revert these changes)

File: modules/ardrivo/test/Arduino.cxx (hunk 30 above)
Comment (AeroStun) [reply modules/ardrivo/test/Arduino.cxx:289]: +1 Indeed `std::array` has lexicographical comparison operators

File: modules/ardrivo/test/HardwareSerial.cxx (hunk 32 above)
Comment (AeroStun) [reply modules/ardrivo/test/HardwareSerial.cxx:59]: Can even be packed into a single `REQUIRE` clause via a `range::equals`

File: modules/ardrivo/test/String.cxx [hunk 36]
```diff
@@ -0,0 +1,215 @@
...
+	auto num = test.toInt();
+	auto num2 = test2.toInt();
+	auto num3 = test3.toInt();
+	auto num4 = test4.toInt();
+	auto num5 = test5.toInt();
+	REQUIRE(num == 1);
+	REQUIRE(num2 == 123);
+	REQUIRE(num3 == 0);
+	REQUIRE(num4 == 0);
+	REQUIRE(num5 == 0);
```
Comment (RuthgerD) [inline modules/ardrivo/test/String.cxx:118]: can be inline `REQUIRE(test.toInt() == 1);`

File: modules/ardrivo/test/String.cxx [hunk 37]
```diff
@@ -0,0 +1,215 @@
...
+	String test = "i like bigger bananas";
+	test.toUpperCase();
+	REQUIRE_THAT(test.c_str(), Catch::Contains("I LIKE BIGGER BANANAS"));
+}
+
+TEST_CASE("Remove spaces in a string" "[trim]")
+{
+	String test = "i like bigger bananas";
+	test.trim();
+	REQUIRE_THAT(test.c_str(), Catch::Contains("ilikebiggerbananas"));
```
Comment (RuthgerD) [inline modules/ardrivo/test/String.cxx:155]: Shoudnt trim only delete leading and trailing whitespaces? [arduino ref](https://www.arduino.cc/reference/en/language/variables/data-types/string/functions/trim)
Looking at our impl its deleting all whitespaces :thinking: 
```cpp
void String::trim() { m_u.erase(std::remove(m_u.begin(), m_u.end(), ' '), m_u.end()); }
```

File: modules/ardrivo/test/String.cxx [hunk 38]
```diff
@@ -0,0 +1,215 @@
...
+	bool isEqual = test.equals(test2);
+	REQUIRE(!isEqual);
+}
+
+TEST_CASE("Checks if a string is equal to another and ignores case" "[equalsIgnoreCase]")
+{
+	String test = "i like bigger bananas";
+	String test2 = "i LIKE BIGGER bananas";
+	bool equalsIgnoreCase = test.equalsIgnoreCase(test2);
+	REQUIRE(equalsIgnoreCase);
```
Comment (RuthgerD) [inline modules/ardrivo/test/String.cxx:171]: can be inline

File: modules/ardrivo/test/String.cxx [hunk 39]
```diff
@@ -0,0 +1,215 @@
...
+
+//Crap does not work without std::string
+TEST_CASE("Test if string is given a new capacity" "[reserve]")
+{ 
+	String cap = "";
+	REQUIRE(cap.length() == 0);
+
+	cap.reserve(25);
+	//REQUIRE(cap.length() == 63);
+}
```
Comment (RuthgerD) [inline modules/ardrivo/test/String.cxx:78]: reserve does not increase the size it only reserves the memory space ([std::basic_string::reserve](https://en.cppreference.com/w/cpp/string/basic_string/reserve)).
You should check if capacity is at least what you reserved -> [https://en.cppreference.com/w/cpp/string/basic_string/capacity](https://en.cppreference.com/w/cpp/string/basic_string/capacity)

File: modules/ardrivo/test/HardwareSerial.cxx [hunk 40]
```diff
@@ -0,0 +1,88 @@
...
+    HardwareSerial test;
+    REQUIRE(test.available() == 3);
+}
+
+TEST_CASE("Check how many avaiable bytes for writing" "[availableForWrite]")
+{
+    init_fake();
+    HardwareSerial test;
+    REQUIRE(test.availableForWrite() == 2147483647);
+}
```
Comment (RuthgerD) [inline modules/ardrivo/test/HardwareSerial.cxx:70]: This is an implementation detail iirc, testing if its available for write at all is enough

File: modules/ardrivo/test/String.cxx (hunk 37 above)
Comment (RuthgerD) [reply modules/ardrivo/test/String.cxx:155]: Correct test case would be:
```cpp
	String test = "    i like bigger bananas    ";
	test.trim();
	REQUIRE_THAT(test == "i like bigger bananas");
```

You have to write tests for whats supposed to be right not what actually works, otherwise you start testing if bugs work correctly

File: modules/ardrivo/test/Arduino.cxx [hunk 41]
```diff
@@ -0,0 +1,329 @@
...
+#include "Arduino.h"
+
+using namespace std::literals;
+
+void static init_fake() 
+{
+    if (board_data)
+        (delete board_data), board_data = nullptr;
+    if (board_info)
+        (delete board_info), board_data = nullptr;
```
Comment (RuthgerD) [inline modules/ardrivo/test/Arduino.cxx:19]: board_data has to be board_info

File: modules/ardrivo/test/Arduino.cxx [hunk 42]
```diff
@@ -0,0 +1,329 @@
...
+    REQUIRE(board_data->pin_frequency[i] == 30);
+}
+
+TEST_CASE("Shifts in a byte of data one bit at a time, start from either the leftmost or rightmost significant bit" "[shiftIn]")
+{
+}
+
+TEST_CASE("Shifts out a byte of data one bit at a time, start from either the leftmost or rightmost significant bit" "[shiftOut]" )
+{
+}
```
Comment (RuthgerD) [inline modules/ardrivo/test/Arduino.cxx:122]: just completely remove them if they do nothing

File: modules/ardrivo/test/Arduino.cxx [hunk 43]
```diff
@@ -0,0 +1,329 @@
...
+    randomSeed(3);
+    std::generate(arr2.begin(), arr2.end(), [](){ return random(100); });
+
+    for (int i = 0; i < 100; i++)
+    {
+        if (arr == arr2)
+        {
+            count++;
+        }
+    }
```
Comment (RuthgerD) [inline modules/ardrivo/test/Arduino.cxx:287]: reformat entire file please
